pkg/checks/external/ssl_util/main.go

```diff
@@ -17,37 +17,80 @@ import (
 
 var TimeoutSeconds = 10
 
-// CertHandshake performs a basic TLS/SSL handshake on the specified host/port, returning any errors
-func CertHandshake(host, port string) error {
-	log.Info("Testing SSL handshake on host ", host, " over port ", port)
+// CertificatePuller makes an insecure connection to a host and specified port, and returns the SSL certificate if the DNSNames field matches the hostname
+func CertificatePuller(host, port string) (*x509.Certificate, error) {
+	var hostCert *x509.Certificate
+	d := &net.Dialer{
+		Timeout: time.Duration(TimeoutSeconds) * time.Second,
+	}
+
+	conn, err := tls.DialWithDialer(d, "tcp", host+":"+port, &tls.Config{
+		InsecureSkipVerify: true,
+		MinVersion:         tls.VersionTLS12,
+	})
+
+	if err != nil {
+		log.Warnln([]*x509.Certificate{&x509.Certificate{}}, "", err)
+		return hostCert, err
+	}
+
+	defer conn.Close()
+
+	cert := conn.ConnectionState().PeerCertificates
+
+	for _, clientCert := range cert {
+		for _, certDNS := range clientCert.DNSNames {
+			if certDNS == host {
+				hostCert = clientCert
+			}
+		}
+	}
+	return hostCert, err
+}
+
+// HandshakeFromHost performs a TLS handshake using a self signed cert, pulled from the host by the CertificatePuller function, returning any errors
+func HandshakeFromHost(host, port string, cert *x509.Certificate) error {
 	d := &net.Dialer{
 		Timeout: time.Duration(TimeoutSeconds) * time.Second,
 	}
 
+	// read the system cert pool, proceed with a new, empty pool if no certs found
+	rootCAs, _ := x509.SystemCertPool()
+	if rootCAs == nil {
+		rootCAs = x509.NewCertPool()
+	}
+
+	if cert != nil {
+		rootCAs.AddCert(cert)
+	}
```

Comment:
and the cert gets checked here whether or not its empty... can we do that check outside of the function? like the check after you call the `CertificatePuller` to see if the result is empty.

Reply:
what happens if it is empty?

Reply:
Updated to error out if an empty certificate is pulled from the host.